iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java

```diff
@@ -0,0 +1,287 @@
+import static org.sonar.iac.docker.checks.utils.command.CommandPredicate.Type.OPTIONAL;
+import static org.sonar.iac.docker.checks.utils.command.CommandPredicate.Type.ZERO_OR_MORE;
+
+public class CommandMatcher {
+
+  private final Deque<ArgumentResolution> argumentStack;
+  private final List<CommandPredicate> detectorPredicates;
+  private Status status;
+  private Deque<CommandPredicate> predicatesStack;
+  private List<ArgumentResolution> argumentsToReport;
+  private CommandPredicate currentPredicate;
+
+  public CommandMatcher(Deque<ArgumentResolution> argumentStack, List<CommandPredicate> detectorPredicates) {
+    this.argumentStack = argumentStack;
+    this.detectorPredicates = detectorPredicates;
+    this.status = Status.CONTINUE;
+  }
+
+  /**
+   * Process and reduce the stack of arguments. Within the loop, which iterates over a stack of predicates,
+   * each argument from the stack is consumed and tested to see if the corresponding predicate is a match.
+   * Each consumed argument that matches is added to the list of arguments that will later form the suitable command.
+   * If a predicate is not optional and does not match, an empty list is returned.
+   * The method is then called again with a reduced argument stack until there are no more arguments on the stack.
+   * If a predicate can be applied multiple times to the argument stack, it is placed on the predicate stack again at the end of the loop.
+   */
+  // Cognitive Complexity of methods should not be too high
+  @SuppressWarnings("java:S3776")
+  public List<ArgumentResolution> fullMatch() {
+    predicatesStack = new LinkedList<>(detectorPredicates);
+    argumentsToReport = new ArrayList<>();
+
+    while (!predicatesStack.isEmpty()) {
+
+      currentPredicate = predicatesStack.pollFirst();
+
+      // resolution is removed from stack during match-methods, here it is only peeked to see if it is null or UNRESOLVED
+      ArgumentResolution resolution = peekFirstArgumentResolution();
+
+      // Stop argument detection when argument list is empty
+      if (resolution == null) {
+        return remainingPredictsAreOptional() ? argumentsToReport : Collections.emptyList();
+      }
+
+      // Stop argument detection when argument is unresolved to start new command detection
+      if (resolution.isUnresolved()) {
+        // remove first element from stack as it is UNRESOLVED
+        pollFirstArgumentResolution();
+        return Collections.emptyList();
+      }
+
+      // choose match-method depending on implementation of currentPredicate
+      if (currentPredicate instanceof SingularPredicate) {
+        matchPredicate((SingularPredicate) currentPredicate);
+      } else if (currentPredicate instanceof OptionPredicate) {
+        matchPredicate((OptionPredicate) currentPredicate);
+      } else if (currentPredicate instanceof MultipleUnorderedOptionsPredicate) {
+        matchPredicate((MultipleUnorderedOptionsPredicate) currentPredicate);
+      }
```

Comment:
Seems a bit strange to me, I understand that we need a different logic to be applied depending of the predicate object, but this goes a bit against the OOP concept.

It would be great to be able to extract the specific logics into the related predicate class.

My default reflex would be to have the interface `CommandPredicate` define a method `match(...)` from which each inheriting class `SingularPredicate`/`OptionPredicate`/`MultipleUnorderedOptionsPredicate` would have an implementation. This implementation would be the current code you have in the different `matchPredicate()` method.
But this cannot be that simple since those depends heavily depend of the `CommandMatcher` context.

Reply:
Sounds like a good and clean approach.

Reply:
How would you suggest handling the context of the `CommandMatcher`? The original idea was to extract the methods and arguments to the `CommandMatcher` to not pass 4 arguments in each method call.

Reply:
One option would be to pass an object to the methods, which wraps the other arguments and the status.
(Essentially an object of the class `CommandMatcher`, which doesn't have the `match()` methods and only works as a wrapper.
But that would be pretty similar to my first approach, where everything was handled in the CommandDetector, and I had an internal class wrapping the arguments. Except that now the `match()` methods would be contained in the implementations of `CommandPredicate` (which I think is a step in the right direction, but not quite ideal)